lib/checks/aria/aria-allowed-attr-evaluate.js

```diff
@@ -27,31 +33,80 @@ import { getRole, allowedAttr, validateAttr } from '../../commons/aria';
  */
 function ariaAllowedAttrEvaluate(node, options, virtualNode) {
   const invalid = [];
-
+  const rowAttrInvalid = [];
   const role = getRole(virtualNode);
   const attrs = virtualNode.attrNames;
   let allowed = allowedAttr(role);
-
   // @deprecated: allowed attr options to pass more attrs.
   // configure the standards spec instead
   if (Array.isArray(options[role])) {
     allowed = uniqueArray(options[role].concat(allowed));
   }
 
+  let tableRole;
+  function validateRowAttrs(validValue) {
+    // check if the parent exists otherwise a TypeError will occur (virtual-nodes specifically)
+    if (virtualNode.parent && role === 'row' && validValue) {
+      // TODO: look into memoizing getRole and removing this local cache
+      let tableMap = cache.get('aria-allowed-attr-table');
+      if (!tableMap) {
+        tableMap = new WeakMap();
+        cache.set('aria-allowed-attr-table', tableMap);
+      }
+
+      const table = closest(
+        virtualNode,
+        'table, [role="treegrid"], [role="table"], [role="grid"]'
+      );
+      tableRole = tableMap.get(table);
+      if (table && !tableRole) {
+        tableRole = getRole(table);
+        tableMap.set(table, tableRole);
+      }
+
+      if (['table', 'grid'].includes(tableRole) && role === 'row') {
+        return true;
+      }
+    }
+  }
+
+  const preChecks = {
+    'aria-posinset': validateRowAttrs,
+    'aria-setsize': validateRowAttrs,
+    'aria-expanded': validateRowAttrs,
+    'aria-level': validateRowAttrs
+  };
+
   if (allowed) {
+    let validValue;
     for (let i = 0; i < attrs.length; i++) {
       const attrName = attrs[i];
+      validValue = validateAttrValue(virtualNode, attrName);
+      if (
+        preChecks[attrName]
+          ? preChecks[attrName](validValue)
+          : false && validateAttr(attrName) && allowed.includes(attrName)
+      ) {
+        rowAttrInvalid.push(attrName + '="' + virtualNode.attr(attrName) + '"');
+      }
       if (validateAttr(attrName) && !allowed.includes(attrName)) {
         invalid.push(attrName + '="' + virtualNode.attr(attrName) + '"');
       }
     }
   }
 
+  if (rowAttrInvalid.length) {
```

Comment:
I don't think we should ignore the other `invalid` array like that. Not sure if that means tweaking this, or adding a check.

Reply:
I have done some clean up using the existing `invalid` array and still checks the aria-attributes